File: ios/Sources/UWBCompassCore/MockRangingProvider.swift
import Foundation

/// A scripted relative trajectory of the peer.
public struct MockPoint: Sendable {
    public let distanceMeters: Double
    public let azimuthDeg: Double
    public init(distanceMeters: Double, azimuthDeg: Double) {
        self.distanceMeters = distanceMeters
        self.azimuthDeg = azimuthDeg
    }
}

public struct MockScript: Sendable {
    public let points: [MockPoint]
    public let lossWindowsMs: [ClosedRange<Int64>]

    public init(points: [MockPoint], lossWindowsMs: [ClosedRange<Int64>] = []) {
        self.points = points
        self.lossWindowsMs = lossWindowsMs
    }

    /// Peer starts 20 m away at +60° and walks straight toward the device.
    public static func approaching(steps: Int = 50) -> MockScript {
        let pts = (0..<steps).map { i -> MockPoint in
            let frac = Double(i) / Double(steps - 1)
            return MockPoint(distanceMeters: 20 * (1 - frac) + 0.5, azimuthDeg: 60 * (1 - frac))
        }
        return MockScript(points: pts)
    }

    /// Approaching, but UWB drops out for a mid window (drives downgrade tests).
    public static func approachingWithDropout(steps: Int = 50) -> MockScript {
        MockScript(points: approaching(steps: steps).points, lossWindowsMs: [1500...2500])
    }

    /// Peer circles the device at a fixed radius.
    public static func circling(steps: Int = 60, radius: Double = 5) -> MockScript {
        let pts = (0..<steps).map { i -> MockPoint in
            MockPoint(distanceMeters: radius, azimuthDeg: wrap180(Double(i) / Double(steps) * 360))
        }
        return MockScript(points: pts)
    }
}

/// Deterministic simulated ranging provider — the exact mirror of the Android mock. Drives
/// all hardware-free tests and SwiftUI previews.
public struct MockRangingProvider: RangingProvider {
    public let technology: Technology
    private let script: MockScript
    private let sampleIntervalMs: Int64
    private let distanceStd: Double
    private let azimuthStd: Double
    private let seed: UInt64

    public init(
        technology: Technology = .UWB,
        script: MockScript = .approaching(),
        sampleIntervalMs: Int64 = 100,
        distanceStd: Double = 0.05,
        azimuthStd: Double = 2.0,
        seed: UInt64 = 42
    ) {
        self.technology = technology
        self.script = script
        self.sampleIntervalMs = sampleIntervalMs
        self.distanceStd = distanceStd
        self.azimuthStd = azimuthStd
        self.seed = seed
    }

    public func samples() -> AsyncStream<RangingSample> {
        AsyncStream { continuation in
            let task = Task {
                var rng = SplitMix64(seed: seed)
                var t: Int64 = 0
                for point in script.points {
                    if Task.isCancelled { break }
                    let lost = script.lossWindowsMs.contains { $0.contains(t) }
                    let sample: RangingSample
                    if lost {
                        sample = RangingSample(technology: technology, distanceMeters: nil,
                                               azimuthDeg: nil, elevationDeg: nil,
                                               quality: .lost, timestampMs: t)
                    } else {
                        let dist = max(0, point.distanceMeters + rng.gaussian() * distanceStd)
                        let az = technology == .UWB
                            ? wrap180(point.azimuthDeg + rng.gaussian() * azimuthStd)
                            : nil
                        sample = RangingSample(technology: technology, distanceMeters: dist,
                                               azimuthDeg: az, elevationDeg: nil,
                                               quality: MockRangingProvider.quality(for: point.distanceMeters, lost: false, tech: technology),
                                               timestampMs: t)
                    }
                    continuation.yield(sample)
                    t += sampleIntervalMs
                    try? await Task.sleep(nanoseconds: UInt64(sampleIntervalMs) * 1_000_000)
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    public static func quality(for distanceMeters: Double, lost: Bool, tech: Technology) -> SignalQuality {
        if lost { return .lost }
        let hi: Double, mid: Double
        switch tech {
        case .UWB: (hi, mid) = (15, 40)
        case .BLE: (hi, mid) = (3, 10)
        case .GPS: (hi, mid) = (20, 100)
        }
        if distanceMeters <= hi { return .high }
        if distanceMeters <= mid { return .medium }
        return .low
    }
}

/// Small deterministic PRNG so simulated runs are reproducible across platforms.
struct SplitMix64 {
    private var state: UInt64
    init(seed: UInt64) { state = seed }

    mutating func next() -> UInt64 {
        state &+= 0x9E3779B97F4A7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58476D1CE4E5B9
        z = (z ^ (z >> 27)) &* 0x94D049BB133111EB
        return z ^ (z >> 31)
    }

    mutating func nextDouble() -> Double {
        Double(next() >> 11) * (1.0 / 9007199254740992.0)
    }

    /// Box-Muller standard normal.
    mutating func gaussian() -> Double {
        let u1 = max(nextDouble(), 1e-9)
        let u2 = nextDouble()
        return (-2 * Foundation.log(u1)).squareRoot() * Foundation.cos(2 * Double.pi * u2)
    }
}

File: android/core/src/main/kotlin/com/uwbcompass/core/MockRangingProvider.kt
package com.uwbcompass.core

import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.random.Random

/**
 * Deterministic simulated ranging provider. Drives all hardware-free tests and Compose
 * previews. It models a peer moving along a scripted path relative to the device and
 * adds configurable Gaussian noise plus optional signal-loss windows.
 */
class MockRangingProvider(
    override val technology: Technology = Technology.UWB,
    private val script: MockScript = MockScript.approaching(),
    private val sampleIntervalMs: Long = 100,
    private val noise: Noise = Noise(),
    private val seed: Long = 42,
) : RangingProvider {

    data class Noise(val distanceStdMeters: Double = 0.05, val azimuthStdDeg: Double = 2.0)

    override fun samples(): Flow<RangingSample> = flow {
        val rng = Random(seed)
        var t = 0L
        for (point in script.points) {
            val lost = script.lossWindows.any { t in it }
            val quality = qualityFor(point.distanceMeters, lost, technology)
            val sample = if (lost) {
                RangingSample(technology, null, null, null, SignalQuality.LOST, t)
            } else {
                val dist = (point.distanceMeters + gaussian(rng) * noise.distanceStdMeters)
                    .coerceAtLeast(0.0)
                // Direction is only available for UWB (ADR-0003).
                val az = if (technology == Technology.UWB) {
                    wrap180(point.azimuthDeg + gaussian(rng) * noise.azimuthStdDeg)
                } else {
                    null
                }
                RangingSample(technology, dist, az, null, quality, t)
            }
            emit(sample)
            t += sampleIntervalMs
            delay(sampleIntervalMs)
        }
    }

    private fun gaussian(rng: Random): Double {
        // Box-Muller.
        val u1 = rng.nextDouble().coerceAtLeast(1e-9)
        val u2 = rng.nextDouble()
        return kotlin.math.sqrt(-2.0 * kotlin.math.ln(u1)) * kotlin.math.cos(2.0 * Math.PI * u2)
    }

    companion object {
        fun qualityFor(distanceMeters: Double, lost: Boolean, tech: Technology): SignalQuality {
            if (lost) return SignalQuality.LOST
            // UWB stays HIGH far longer than BLE; GPS is coarse.
            val (hi, mid) = when (tech) {
                Technology.UWB -> 15.0 to 40.0
                Technology.BLE -> 3.0 to 10.0
                Technology.GPS -> 20.0 to 100.0
            }
            return when {
                distanceMeters <= hi -> SignalQuality.HIGH
                distanceMeters <= mid -> SignalQuality.MEDIUM
                else -> SignalQuality.LOW
            }
        }
    }
}

/** A scripted relative trajectory of the peer. */
data class MockPoint(val distanceMeters: Double, val azimuthDeg: Double)

data class MockScript(
    val points: List<MockPoint>,
    val lossWindows: List<LongRange> = emptyList(),
) {
    companion object {
        /** Peer starts 20 m away at +60° and walks straight toward the device. */
        fun approaching(steps: Int = 50): MockScript {
            val pts = (0 until steps).map { i ->
                val frac = i.toDouble() / (steps - 1)
                MockPoint(distanceMeters = 20.0 * (1 - frac) + 0.5, azimuthDeg = 60.0 * (1 - frac))
            }
            return MockScript(pts)
        }

        /** Approaching, but UWB drops out for a mid window (drives downgrade tests). */
        fun approachingWithDropout(steps: Int = 50): MockScript {
            val base = approaching(steps)
            return base.copy(lossWindows = listOf(1500L..2500L))
        }

        /** Peer circles the device at a fixed radius. */
        fun circling(steps: Int = 60, radius: Double = 5.0): MockScript {
            val pts = (0 until steps).map { i ->
                val ang = wrap180(i.toDouble() / steps * 360.0)
                MockPoint(radius, ang)
            }
            return MockScript(pts)
        }
    }
}

/** Wrap an angle into (-180, 180]. Shared with fusion math. */
fun wrap180(deg: Double): Double {
    var a = deg % 360.0
    if (a > 180.0) a -= 360.0
    if (a <= -180.0) a += 360.0
    return a
}

internal fun angularDistance(a: Double, b: Double): Double = abs(wrap180(a - b))

/** Smallest signed rotation from [from] to [to], in (-180, 180]. */
fun signedAngleDelta(from: Double, to: Double): Double {
    val d = wrap180(to - from)
    return d
}

internal fun bearingFromComponents(x: Double, y: Double): Double =
    Math.toDegrees(atan2(y, x))
